File: ios/mixBase/Services/MixbaseAPI.swift
import Foundation

// MARK: - MixbaseAPI
// Client for the web app's authenticated API routes (mixbase.app). These are
// the routes that run the paid AI generation server-side — artwork via
// Replicate, visualizers via Runway — with per-tier limits enforced where the
// keys live. The middleware accepts `Authorization: Bearer <supabase access
// token>`, which is exactly how this client authenticates.

final class MixbaseAPI {

    static let shared = MixbaseAPI()

    private let baseURL = Config.apiBaseURL

    // Generation routes block while the server polls the AI provider — artwork
    // up to 2 min, Runway video up to 5 min — so the session must allow far
    // more than URLSession's default 60s per-request timeout.
    private let session: URLSession

    private let decoder: JSONDecoder

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 6 * 60
        config.timeoutIntervalForResource = 8 * 60
        self.session = URLSession(configuration: config)

        // Same tolerant date handling as SupabaseService: ISO 8601 with and
        // without fractional seconds (Next.js/PostgREST emit both).
        let isoFractional = ISO8601DateFormatter()
        isoFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let isoPlain = ISO8601DateFormatter()
        isoPlain.formatOptions = [.withInternetDateTime]

        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let dateString = try container.decode(String.self)
            if let date = isoFractional.date(from: dateString) { return date }
            if let date = isoPlain.date(from: dateString) { return date }
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Cannot decode date: \(dateString)"
            )
        }
    }

    // MARK: - Image models
    // Mirrors IMAGE_MODELS in src/lib/artwork-models.ts — ids must match the
    // server registry. First entry is the default.
    struct ImageModel: Identifiable {
        let id: String
        let label: String
    }

    static let imageModels: [ImageModel] = [
        ImageModel(id: "flux-ultra",   label: "FLUX Ultra Raw"),
        ImageModel(id: "seedream",     label: "Seedream 4"),
        ImageModel(id: "imagen-ultra", label: "Imagen 4 Ultra"),
        ImageModel(id: "recraft",      label: "Recraft V3"),
        ImageModel(id: "flux",         label: "Flux 2 Pro"),
        ImageModel(id: "imagen",       label: "Imagen 4"),
    ]

    // MARK: - Artwork

    /// Generate AI artwork for a project. The server generates the image,
    /// uploads it to storage AND applies it as the project's artwork.
    /// Returns the public URL of the applied artwork.
    func generateArtwork(projectId: UUID, prompt: String, model: String, vary: Bool) async throws -> String {
        let body: [String: Any] = [
            "project_id": projectId.uuidString.lowercased(),
            "prompt": prompt,
            "model": model,
            "vary": vary,
        ]
        let json = try await requestJSON(path: "/api/generate-artwork", method: "POST", body: body)
        guard let url = json["artwork_url"] as? String else {
            throw MixbaseAPIError.invalidResponse("No artwork URL in response")
        }
        return url
    }

    // MARK: - Visualizers

    /// Available Runway image-to-video models with their valid durations/ratios.
    func fetchRunwayModels() async throws -> [RunwayModel] {
        let data = try await requestData(path: "/api/visualizer/runway", method: "GET")
        struct ModelsResponse: Codable { let models: [RunwayModel] }
        return try decoder.decode(ModelsResponse.self, from: data).models
    }

    /// Generate a visualizer video from an artwork image. Blocks until the
    /// server finishes polling Runway (can take minutes for slower models).
    /// The server persists the video to the user's Media library when it can.
    /// Returns the video URL (a permanent storage URL when saved=true).
    func generateVisualizer(
        projectId: UUID,
        imageUrl: String,
        model: String,
        duration: Int,
        ratio: String,
        prompt: String?
    ) async throws -> (videoUrl: String, saved: Bool) {
        var body: [String: Any] = [
            "projectId": projectId.uuidString.lowercased(),
            "imageUrl": imageUrl,
            "model": model,
            "duration": duration,
            "ratio": ratio,
        ]
        if let prompt, !prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            body["promptText"] = prompt
        }
        let json = try await requestJSON(path: "/api/visualizer/runway", method: "POST", body: body)
        guard let url = json["videoUrl"] as? String else {
            throw MixbaseAPIError.invalidResponse("No video URL in response")
        }
        return (url, (json["saved"] as? Bool) ?? false)
    }

    /// Every saved visualizer the user owns, newest first.
    func fetchVisualizers() async throws -> [Visualizer] {
        let data = try await requestData(path: "/api/visualizer", method: "GET")
        return try decoder.decode([Visualizer].self, from: data)
    }

    /// Delete a saved visualizer (also un-pins it from any project server-side).
    func deleteVisualizer(id: UUID) async throws {
        _ = try await requestData(path: "/api/visualizer/\(id.uuidString.lowercased())", method: "DELETE")
    }

    /// Pin (or clear, with nil) a video as a project's visualizer. The server
    /// verifies the URL is a visualizer the user actually owns.
    func pinVisualizer(projectId: UUID, videoUrl: String?) async throws {
        let body: [String: Any] = ["visualizer_url": videoUrl ?? NSNull()]
        _ = try await requestJSON(path: "/api/projects/\(projectId.uuidString.lowercased())", method: "PATCH", body: body)
    }

    // MARK: - Core request plumbing

    /// Perform a request and parse the response as a JSON object.
    private func requestJSON(path: String, method: String, body: [String: Any]? = nil) async throws -> [String: Any] {
        let data = try await requestData(path: path, method: method, body: body)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw MixbaseAPIError.invalidResponse("Response was not a JSON object")
        }
        return json
    }

    /// Perform an authenticated request. On 401, refreshes the Supabase session
    /// (coalesced in AuthService) and retries once with the new token. Non-2xx
    /// responses surface the server's own `error` message — that's where the
    /// tier-limit and upgrade prompts come from.
    private func requestData(path: String, method: String, body: [String: Any]? = nil) async throws -> Data {
        func makeRequest(token: String?) throws -> URLRequest {
            guard let url = URL(string: "\(baseURL)\(path)") else {
                throw MixbaseAPIError.invalidResponse("Bad URL: \(path)")
            }
            var request = URLRequest(url: url)
            request.httpMethod = method
            if let token {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            if let body {
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.httpBody = try JSONSerialization.data(withJSONObject: body)
            }
            return request
        }

        var (data, response) = try await session.data(for: makeRequest(token: currentToken()))

        // Expired access token — refresh once and retry with the new one.
        if let http = response as? HTTPURLResponse, http.statusCode == 401 {
            let refreshed = await AuthService.shared.refreshSession()
            guard refreshed else { throw MixbaseAPIError.notAuthenticated }
            (data, response) = try await session.data(for: makeRequest(token: currentToken()))
        }

        guard let http = response as? HTTPURLResponse else {
            throw MixbaseAPIError.invalidResponse("Not an HTTP response")
        }
        guard (200...299).contains(http.statusCode) else {
            // Prefer the server's human-readable error (e.g. "Monthly artwork
            // limit reached (3/3). Upgrade to generate more.")
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let message = json["error"] as? String {
                throw MixbaseAPIError.serverError(message)
            }
            throw MixbaseAPIError.httpError(statusCode: http.statusCode)
        }
        return data
    }

    /// The current Supabase access token, as persisted by AuthService.
    private func currentToken() -> String? {
        KeychainService.load(forKey: "access_token")
    }
}

// MARK: - MixbaseAPIError

enum MixbaseAPIError: LocalizedError {
    case notAuthenticated
    case serverError(String)
    case httpError(statusCode: Int)
    case invalidResponse(String)

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Your session expired. Please sign in again."
        case .serverError(let message):
            return message
        case .httpError(let code):
            return "Request failed (HTTP \(code))"
        case .invalidResponse(let message):
            return message
        }
    }
}

File: android/core/src/main/kotlin/app/mixbase/core/api/MixbaseApi.kt
package app.mixbase.core.api

import app.mixbase.core.Config
import app.mixbase.core.MixbaseException
import app.mixbase.core.auth.SessionManager
import app.mixbase.core.json.AppJson
import app.mixbase.core.model.FeedComment
import app.mixbase.core.model.FeedItem
import app.mixbase.core.model.Feedback
import app.mixbase.core.model.Version
import app.mixbase.core.model.Visualizer
import app.mixbase.core.net.HttpResult
import app.mixbase.core.net.JSON_MEDIA_TYPE
import app.mixbase.core.net.awaitResult
import app.mixbase.core.net.forLongRequests
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

/**
 * Client for the web app's authenticated API routes (mixbase.app). These
 * run the paid AI generation server-side (Replicate) with the monthly
 * allowance enforced where the keys live, serve the cross-user community
 * feed, and own the writes that need server judgement. The middleware
 * accepts `Authorization: Bearer <supabase access token>`.
 *
 * Port of ios/mixBase/Services/MixbaseAPI.swift.
 */
class MixbaseApi(
    http: OkHttpClient,
    private val session: SessionManager,
    private val baseUrl: String = Config.API_BASE_URL,
) {
    private val http: OkHttpClient = http.forLongRequests()

    data class ImageModel(val id: String, val label: String)

    // ── Artwork ─────────────────────────────────────────────────────────────

    /**
     * Generate AI artwork for a project. The server generates the image,
     * uploads it to storage AND applies it as the project's artwork.
     * Returns the public URL of the applied artwork.
     */
    suspend fun generateArtwork(projectId: String, prompt: String, model: String, vary: Boolean): String {
        val json = requestJson("/api/generate-artwork", "POST", buildJsonObject {
            put("project_id", projectId)
            put("prompt", prompt)
            put("model", model)
            put("vary", vary)
        })
        return json["artwork_url"]?.jsonPrimitive?.contentOrNull
            ?: throw MixbaseException.InvalidResponse("No artwork URL in response")
    }

    /** Set an existing artwork image as a project's cover (must be a Supabase storage URL). */
    suspend fun assignArtworkToProject(projectId: String, artworkUrl: String) {
        requestJson("/api/projects/$projectId", "PATCH", buildJsonObject { put("artwork_url", artworkUrl) })
    }

    // ── Visualizers (view / pin / delete only — generation is web-only) ────

    suspend fun fetchVisualizers(): List<Visualizer> =
        decode(ListSerializer(Visualizer.serializer()), requestData("/api/visualizer", "GET"))

    suspend fun deleteVisualizer(id: String) { requestData("/api/visualizer/$id", "DELETE") }

    /** Pin (or clear, with null) a video as a project's visualizer. */
    suspend fun pinVisualizer(projectId: String, videoUrl: String?) {
        requestJson("/api/projects/$projectId", "PATCH", buildJsonObject {
            if (videoUrl == null) put("visualizer_url", JsonNull) else put("visualizer_url", videoUrl)
        })
    }

    // ── Versions ────────────────────────────────────────────────────────────

    /**
     * Create the mb_versions row for a mix that has just finished uploading.
     *
     * Goes through the web route rather than PostgREST because three columns
     * are the SERVER's decision: `allow_download` (a consent signal the route
     * inherits from the artist's previous choice ONLY when absent — never send
     * it), `status`/`label` (parsed from the filename), and `version_number`
     * (max+1, retried on conflict). No fallback to a direct insert.
     */
    suspend fun createVersion(
        projectId: String,
        audioUrl: String,
        audioFilename: String? = null,
        durationSeconds: Int? = null,
        fileSizeBytes: Long? = null,
    ): Version {
        val body = buildJsonObject {
            put("project_id", projectId)
            put("audio_url", audioUrl)
            // Omit rather than send null: the route forwards these into the insert.
            if (audioFilename != null) put("audio_filename", audioFilename)
            if (durationSeconds != null) put("duration_seconds", durationSeconds)
            if (fileSizeBytes != null) put("file_size_bytes", fileSizeBytes)
        }
        return decode(Version.serializer(), requestData("/api/versions", "POST", body))
    }

    /** Share-link download consent — a signal, not an access control (src/lib/version-defaults.ts). */
    suspend fun setAllowDownload(versionId: String, allow: Boolean) {
        requestJson("/api/versions/$versionId", "PATCH", buildJsonObject { put("allow_download", allow) })
    }

    /** Owner-only timestamped note on one of your own mixes (same route as the web player). */
    suspend fun postMixNote(versionId: String, comment: String, timestampSeconds: Int): Feedback =
        decode(Feedback.serializer(), requestData("/api/mix-notes", "POST", buildJsonObject {
            put("version_id", versionId)
            put("comment", comment)
            put("timestamp_seconds", timestampSeconds)
        }))

    // ── Community feed (cross-user by design) ───────────────────────────────

    /** Recent uploads across ALL artists. One malformed row is dropped, not the whole feed. */
    suspend fun fetchFeed(): List<FeedItem> {
        val result = requestData("/api/feed", "GET")
        val array = runCatching { AppJson.parseToJsonElement(result.body).jsonArray }.getOrNull()
            ?: throw MixbaseException.InvalidResponse("Couldn't read the feed")
        return array.mapNotNull { element ->
            runCatching { AppJson.decodeFromJsonElement(FeedItem.serializer(), element) }.getOrNull()
        }
    }

    suspend fun postFeedComment(versionId: String, comment: String): FeedComment =
        decode(FeedComment.serializer(), requestData("/api/feed/comments", "POST", buildJsonObject {
            put("version_id", versionId)
            put("comment", comment)
        }))

    /** Report objectionable feed content; type is "version" or "comment". */
    suspend fun reportContent(type: String, id: String, reason: String? = null) {
        requestData("/api/feed/report", "POST", buildJsonObject {
            put("content_type", type)
            put("content_id", id)
            if (!reason.isNullOrBlank()) put("reason", reason)
        })
    }

    /** Block another artist — their uploads and comments disappear from this account's feed. */
    suspend fun blockUser(userId: String) {
        requestData("/api/feed/block", "POST", buildJsonObject { put("user_id", userId) })
    }

    // ── Account ─────────────────────────────────────────────────────────────

    /** Permanently delete the signed-in account and all its data. */
    suspend fun deleteAccount() { requestData("/api/auth/delete-account", "POST") }

    // ── Plumbing ────────────────────────────────────────────────────────────

    private suspend fun requestJson(path: String, method: String, body: JsonObject? = null): JsonObject {
        val result = requestData(path, method, body)
        return runCatching { AppJson.parseToJsonElement(result.body).jsonObject }.getOrNull()
            ?: throw MixbaseException.InvalidResponse("Response was not a JSON object")
    }

    private fun <T> decode(serializer: kotlinx.serialization.KSerializer<T>, result: HttpResult): T = try {
        AppJson.decodeFromString(serializer, result.body)
    } catch (e: Exception) {
        throw MixbaseException.InvalidResponse("Couldn't read the server's response")
    }

    /**
     * Authenticated request. On 401, refreshes the Supabase session (coalesced
     * in [SessionManager]) and retries once. Non-2xx surfaces the server's own
     * `error` message.
     */
    internal suspend fun requestData(path: String, method: String, body: JsonObject? = null): HttpResult {
        fun build(token: String?): Request {
            val builder = Request.Builder().url(baseUrl + path)
            if (token != null) builder.header("Authorization", "Bearer $token")
            val requestBody = body?.let {
                builder.header("Content-Type", "application/json")
                it.toString().toRequestBody(JSON_MEDIA_TYPE)
            }
            return builder.method(method, requestBody ?: if (method == "GET") null else "".toRequestBody(null)).build()
        }

        var result = execute(build(session.accessToken))
        if (result.code == 401) {
            if (!session.refreshSession()) throw MixbaseException.NotAuthenticated()
            result = execute(build(session.accessToken))
        }
        if (result.isSuccess) return result
        if (result.code == 401) throw MixbaseException.NotAuthenticated()

        val json = runCatching { AppJson.parseToJsonElement(result.body).jsonObject }.getOrNull()
        if (json != null) {
            // Monthly allowance exhausted. There are no paid plans anywhere in
            // mixBase, so the copy stays purchase-free.
            if (json["upgrade"]?.jsonPrimitive?.booleanOrNull == true) {
                throw MixbaseException.Server(result.code, LIMIT_MESSAGE)
            }
            val message = json["error"]?.jsonPrimitive?.contentOrNull
            if (message != null) throw MixbaseException.Server(result.code, scrubPurchaseWords(message))
        }
        throw MixbaseException.Http(result.code)
    }

    private suspend fun execute(request: Request): HttpResult = try {
        http.newCall(request).awaitResult()
    } catch (e: IOException) {
        throw MixbaseException.Network(e)
    }

    companion object {
        /** Mirrors IMAGE_MODELS in src/lib/artwork-models.ts — ids must match the server registry. */
        val IMAGE_MODELS: List<ImageModel> = listOf(
            ImageModel("flux-ultra", "FLUX Ultra Raw"),
            ImageModel("nano-pro", "Nano Banana Pro"),
            ImageModel("flux-krea", "FLUX Krea"),
            ImageModel("seedream", "Seedream 4"),
            ImageModel("flux", "Flux 2 Pro"),
            ImageModel("nano", "Nano Banana 2"),
        )

        const val LIMIT_MESSAGE = "You've reached this month's limit for AI generations. It resets at the start of next month."
        const val NEUTRAL_MESSAGE = "This action isn't available right now. Please try again later."

        private val purchaseWords = listOf("upgrade", "plan", "tier", "subscri", "purchase", "billing", "pricing", "credit", "buy ")

        /** Never show copy that steers to a purchase — there is nothing to buy. */
        fun scrubPurchaseWords(message: String): String {
            val lowered = message.lowercase()
            return if (purchaseWords.any { lowered.contains(it) }) NEUTRAL_MESSAGE else message
        }
    }
}
